keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -192,584 +174,1368 @@ logs are lost.
+  lifecycle (as oppose to startup and liveness probes). If Container's readiness
+  probe depends on external services, and those services are not yet available
+  or depend on other containers in the Pod, the sidecar container will never be
+  started.
+- it may be harder semantically to understand what does the not Ready Sidecar
+  container mean if it became not ready after it has started, while other Init
+  containers are being run.
+- the scenario when sidecar container is started, but wants to delay the
+  readiness for later would not be possible to implement.
+
+#### Startup probes
+
+The second possible signal we can use is Startup probes. Kubelet will be waiting
+for startup probe to succeed to mark the container as started. Readiness probes
+will only start when Pod's initialization stage will complete and regular
+containers will start. There is a `ContainerStatus` field `Started` indicating
+this state.
+
+#### `postStart` hook
+
+The easiest approach would be to simply wait for postStart hook to complete.
+This is already implemented for regular containers - they are started
+sequentially with the `postStart` hook being a blocking call.
+
+Using `postStart` has some debugging challenges in how logs are reported (they
+aren't) and the status is displayed. If this option will be used, we should
+consider improving this situation.
+
+With `postStart` simple sidecars will start faster than if any probes are used.
+However it brings additional challenges for sidecar authors as majority of logic
+may be duplicated and needs to be authored, e.g. a shell binary + a loop script
+to retry `curl`.
+
+### Alternative 1. Sidecar containers as a separate collection
+
+This alternative was rejected for the following reasons:
+- it introduces a whole new collection of containers, which is a big change
+doesn’t allow different startup sequencing - before or after init containers.
+Before or after a specific init container. Making a single decision for the
+entire collection of containers and all scenarios was not possible.
+- A single semantic for the entire collection of containers may make it harder
+  to introduce new properties to containers in this collection without changing
+  its semantic. It was preferable to introduce new properties representing
+  specific behaviors to the existing containers
+- Each time a new collection of containers is added to Pod, it increases the
+  size of the OpenAPI schema of pods and all resources containing PodTemplate
+  significantly, which has negative downstream implications on systems consuming
+  OpenAPI schemas and on CRDs that embed PodTemplate.
+
+### Alternative 2. DependOn semantic between containers
+
+This alternative is trying to replicate systemd dependencies. The alternative
+was rejected for the following reasons:
+- Universal sidecars injection is not possible for such a system.
+- There were lack of scenarios requiring more than general containers sequencing
```

Comment:
I will try to gather more scenarios - if people hear about them, please send them my way (I may create a separate tracking issue)